graph/patterns/upsert.md

```diff
@@ -0,0 +1,173 @@
+
+## Solution
+
+The solution is to use an `UPSERT` pattern, to solve for the non-idempotent creation and client-provided naming problems.
+
+* For IaC scenarios, resources must use `UPSERT` semantics with an [alternate key](./alternate-key.md):
+  * Use `PATCH` with a client-provided alternate key.
+  * For a non-existent resource (specified by the alternate key) the service must handle this as a "create". As part of creation, the service must still generate the primary key value.
+  * For an existing resource (specified by the alternate key) the service must handle this as an "update.
+  * Any new alternate key, used for IaC scenarios, should be called `uniqueName`, if there isn't already an existing property that could be used as an alternate key.
+* NOTE: the service must also support `GET` using the alternate key pattern.
+* For consistent CRUD Microsoft Graph behaviors, all resources, **including** resources used in IaC scenarios, should use `POST` and a service-generated primary key, per existing guidelines, and support `GET`, `PATCH` and `DELETE` using the primary key.
+* If a service does not support `UPSERT`, then a `PATCH` call against a non-existent resource must result in an HTTP "409 conflict" error.
+
+This solution allows for existing resources that follow Microsoft Graph conventions for CRUD operations to add `UPSERT` without impacting existing apps or functionality.
+
+Ideally, all new entity types should support an `UPSERT` mechanism, especially if the resource is likely be used in IaC scenarios.
+
+## When to use this pattern
+
+This pattern should be adopted for resources that are managed through infrastructure as code or desired state configuration.
+
+## Issues and considerations
+
+* The addition of this new pattern (with alternate key) does not represent a breaking change.
+However, some API producers may have concerns about accidental usages of this new pattern unwittingly creating many new resources when the intent was an update.
+As a result, API producers can use the `Prefer: idempotent` to require clients to opt-in to the UPSERT behavior.
+* The client-provided alternate key must be immutable after being set. If its value is null then it should be settable as a way to backfill existing resources for use in IaC scenarios.
+* API producers could use `PUT` operations to create or update, but generally this approach is not recommended due to the destructive nature of `PUT`'s replace semantics.
+* API producers could to use `UPSERT` with a primary (client-provided) key and this may be appropriate for some scenarios. However, the recommendation is for resources to support creation using `POST` and a service-generated primary key, for consistency reasons.
+* API producers may annotate entity sets, singletons and collections to indicate that entities can be "upserted". The example below shows this annotation for the `groups` entity set.  
```

Comment:
I'm not sure how Upsert will work for a singleton on Graph.

Reply:
Good catch.  I'll need to check this.  It's not like a singleton needs a client provided key.

Reply:
@mikepizzo what is the behaviour for upsert to a singleton?
My guess is that it's still preferable to have a client-provided key in these cases as this allows for referencing in the template, and in things like Azure Deployment Stacks.

However, there'll be some singletons (like a Policy) that are never created, but always exist in the system, and can only be PATCHed, that wouldn't have a client-provided key (really they're keyless singletons).

Reply:
Note that a PATCH to a singleton doesn't use a primary/alternate key in the URL; it's just a regular PATCH to the navigation property. Upsert should apply here as well; if I do a PATCH to users/1/calendar, and the calendar doesn't exist (it is null) it should be created with the provided values, ortherwise it should be updated with the provided values.  As opposed to PUT which replaces the existing entity with the provided values and sets unspecified values to default.

Reply:
@OlgaPodo - do I need to provide another example for singletons, or should I just add a note to the considerations?  I'm leaning towards the latter.